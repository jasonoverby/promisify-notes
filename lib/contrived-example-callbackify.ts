import { callbackify } from 'util';
import { getMsgAfterRandomSecsAsync } from './contrived-example-promisify';

export const getMsgAfterRandomSecsCallback = callbackify(getMsgAfterRandomSecsAsync);
