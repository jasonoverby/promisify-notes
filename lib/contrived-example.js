const { promisify } = require('util');
const {
  getImprovedErrorMsg,
  getNotANumberMsg,
  getRandomIntBetweenMinAndMax,
  getWaitedSecsMsg,
  notANumber,
} = require('../helpers/helpers');

const getMsgAfterRandomSecsWithCallback = (num1, num2, callback) => {
  // will be NaN if either num1 or num2 is not a number
  const randomInt = getRandomIntBetweenMinAndMax(num1, num2);
  const randomWaitFrom1ToNumSecs = notANumber(randomInt) ? 0 : randomInt * 1000;
  // simulate async operation
  setTimeout(() => {
    if (notANumber(randomInt)) {
      callback(TypeError(getNotANumberMsg(num1, num2)), null);
    } else {
      callback(null, getWaitedSecsMsg(randomInt));
    }
  }, randomWaitFrom1ToNumSecs);
};

const getMsgAfterRandomSecsAsync = async (num1, num2) => {
  const funcWithoutCallback = promisify(getMsgAfterRandomSecsWithCallback);
  return funcWithoutCallback(num1, num2)
    .catch(err => getImprovedErrorMsg(err));
};

module.exports = {
  getMsgAfterRandomSecsAsync,
  getMsgAfterRandomSecsWithCallback,
};
