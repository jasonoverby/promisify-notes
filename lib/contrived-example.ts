import {
  getNotANumberMsg,
  getRandomIntBetweenMinAndMax,
  getWaitedSecsMsg,
  notANumber,
} from '../helpers';

type StringCallback = (err?: Error, str?: string) => void;

const getMsgAfterRandomSecsWithCallback = (num1: number, num2: number, callback: StringCallback): void => {
  // will be NaN if either num1 or num2 is not a number
  const randomInt = getRandomIntBetweenMinAndMax(num1, num2);
  const randomWaitFrom1ToNumSecs = notANumber(randomInt) ? 0 : randomInt * 1000;
  // simulate async operation
  setTimeout(() => {
    if (notANumber(randomInt)) {
      callback(TypeError(getNotANumberMsg(num1, num2)), null);
    } else {
      callback(null, getWaitedSecsMsg(randomInt));
    }
  }, randomWaitFrom1ToNumSecs);
};

export {
  getMsgAfterRandomSecsWithCallback,
  StringCallback,
};
