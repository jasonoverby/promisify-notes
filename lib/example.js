

import { promisify } from 'util';



const funcWithCb = callbackify(asyncFunc);
const callback = (err, data) => {
  if (err) throw err;
  doSomethingWithData(data);
};
funcWithCb(arg, callback);

const asyncFunc = promisify(funcWithCb);
const main = async () => {
  const data = await asyncFunc(arg)
    .catch(err => handleError(err));
  doSomethingWithData(data);
};
main();
/** OR */
asyncFunc(arg)
  .then(data => {
    doSomethingWithData(data);
  }).catch(err => handleError(err));


(async () => {
  const result = await promisifiedFunc();
  otherFunc(result);
})();
const funcWithCallback = (param, callback) => {
  doSomethingAsynchronous()
};


get[promisify.custom] = (
  url,
  options = {},
) => (
  new Promise((resolve, reject) => {
    const callback = (response) => {
      let str = '';
      response.on('data', (data) => {
        str += data;
      });
      response.on('end', () => {
        resolve(JSON.parse(str));
      });
    };
    get(url, options, callback)
      .on('error', reject);
  })
);

import { promisify } from 'util';
import { readFile } from 'fs';

const readFileAsync = promisify(readFile);

const callback = (err, data) => {
  if (err) throw err;
  console.log(data);
};
fs.readFile('file.txt', callback);

/** callback is first param */
const hmmFunc = (callback, str) => {
  const waitTime = getNumBetween(1, 4);
  setTimeout(() => {
    /** no error given to callback */
    callback(getMsg(waitTime, str));
  }, waitTime);
};
// const hmmFuncAsync = ???


hmmFunc[promisify.custom] = (str) => (
  new Promise((resolve) => {
    const waitTime = getNumBetween(1, 4);
    setTimeout(() => {
      resolve(getMsg(waitTime, str));
    }, waitTime);
  })
);
const hmmFuncAsync = promisify(hmmFunc);
const msg = await hmmFuncAsync('hello');

export { hmmFunc };


const hmmFuncAsync = promisify(hmmFunc);
const msg = await hmmFuncAsync('Rad!');

const wait = promisify(setTimeout);
/** waits for at least 3 seconds */
await wait(3000); 
/** before moving on to the next line */
doSomethingAfterWaiting();


try {
  const data = await promisifiedFunc(something);
  const newData = doSomethingWithData(data);
  const newerData = await otherPromisifiedFunc(newData);
  const evenNewerData = doSomethingElseWithData(newerData);
  const newestData = await yetAnotherPromisifiedFunc(evenNewerData);
  doSomethingFinalWithData(newestData);
} catch (err) {
  handleError(err);
}

funcWithCallback(something, (err, data) => {
  if (err) handleError(err);
  const newData = doSomethingWithData(data);

  otherFuncWithCallback(newData, (err, data) =>Â {
    if (err) handleError(err);

      const newData = doSomethingElseWithData(data);
      yetAnotherFuncWithCallback(newData, (err, data) => {
        if (err) handleError(err);

        doSomethingFinalWithData(newData);
      });
  });
});

const dataObj = await promisifiedHttpsGet(url)
  .catch(err => handleError(err));
doSomethingWithDataObj(dataObj);

https.get(url, (error, response) => {
  if (error) handleError(error);

  let str = '';
  response.on('data', (data) => {
    str += data;
  });
  response.on('end', () => {
    const dataObj = JSON.parse(str);
    doSomethingWithDataObj(dataObj);
    done();
  });
};

get(`${todosUrl}/${todoId}`, {}, callback);

import { promisify } from 'util';

const promisifiedFunc = promisify(funcWithCallback);


(async () => {
  const result = await promisifiedFunc();
  otherFunc(result);
})();

console.log(result);
import {
  getNotANumberMsg,
  getRandomIntBetweenMinAndMax,
  getWaitedSecsMsg,
  notANumber,
} from '../helpers';
import { Http2SecureServer } from 'http2';

type StringCallback = (err?: Error, str?: string) => void;

const getMsgAfterRandomSecsWithCallback = (
  num1: number,
  num2: number,
  callback: StringCallback,
): void => {
  const randomInt = getRandomIntBetweenMinAndMax(num1, num2);
  const randomWaitFrom1ToNumSecs = notANumber(randomInt) ?
    0 :
    randomInt * 1000;
  setTimeout(() => {
    if (notANumber(randomInt)) {
      callback(TypeError(getNotANumberMsg(num1, num2)), null);
    } else {
      callback(null, getWaitedSecsMsg(randomInt));
    }
  }, randomWaitFrom1ToNumSecs);
};

const getMsgAfterRandomSecsAsync = promisify(getMsgAfterRandomSecsWithCallback);



get[promisify.custom] = (url) => (
  new Promise((resolve, reject) => {
    get(url, (response) => {
      let str = '';
      response.on('data', (data) => {
        str += data;
      });
      response.on('end', () => {
        resolve(JSON.parse(str));
      });
    }).on('error', reject);
  })
);

export {
  getMsgAfterRandomSecsAsync,
  getMsgAfterRandomSecsWithCallback,
  StringCallback,
};
