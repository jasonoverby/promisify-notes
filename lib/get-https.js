const { promisify } = require('util');
const https = require('https');

https.get[promisify.custom] = url => (
  new Promise((resolve, reject) => {
    https.get(url, (response) => {
      let str = '';
      response.setEncoding('utf8');
      response.on('data', (data) => {
        str += data;
      });
      response.on('end', () => {
        resolve(JSON.parse(str));
      });
    }).on('error', reject);
  })
);

const getHttps = promisify(https.get);

module.exports = {
  getHttps,
};
