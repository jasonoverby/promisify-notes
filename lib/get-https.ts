import * as https from 'https';
import { promisify } from 'util';

type GetHttpsType = (url: string, options?: https.RequestOptions, cb?: (res: object) => void) => Promise<{}>;
https.get[promisify.custom] = (url: string): Promise<object> => (
  new Promise((resolve, reject) => {
    https.get(url, (response) => {
      let str = '';
      response.setEncoding('utf8');
      response.on('data', (data) => {
        str += data;
      });
      response.on('end', () => {
        resolve(JSON.parse(str));
      });
    }).on('error', reject);
  })
);

export const getHttps: GetHttpsType = promisify(https.get);
