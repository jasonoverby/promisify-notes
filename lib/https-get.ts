import { IncomingMessage, ServerResponse } from 'http';
import {
  get,
  RequestOptions,
} from 'https';
import { promisify } from 'util';

type HttpsGetCb = (response: IncomingMessage) => void;
type HttpsGetType = (url: string | URL, options?: RequestOptions, callback?: HttpsGetCb) => {};
get[promisify.custom] = (url: string, options: RequestOptions = {}): Promise<{}> => (
  new Promise((resolve, reject) => {
    const callback: HttpsGetCb = (response: IncomingMessage) => {
      let str = '';
      response.on('data', (data: ServerResponse) => {
        str += data;
      });
      response.on('end', () => {
        resolve(JSON.parse(str));
      });
    };
    get(url, options, callback).on('error', reject);
  })
);

export {
  get,
  HttpsGetType,
  HttpsGetCb,
};
