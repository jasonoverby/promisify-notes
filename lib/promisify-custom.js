const { promisify } = require('util');

const getSomethingErr = something => new TypeError(`"${something}" is not something`);

const myFuncWithCallback = (something, callback) => {
  if (something !== 'something') {
    callback(null, getSomethingErr(something));
  } else {
    callback(something, null);
  }
};

const myFuncAsyncThatWillNotRun = promisify(myFuncWithCallback);

myFuncWithCallback[promisify.custom] = something => (
  new Promise((res, rej) => {
    if (something !== 'something') {
      const err = getSomethingErr(something);
      rej(err);
    } else {
      res(`${something} async`);
    }
  })
);

const myFuncAsync = promisify(myFuncWithCallback);

module.exports = {
  myFuncWithCallback,
  myFuncAsync,
  myFuncAsyncThatWillNotRun,
};
