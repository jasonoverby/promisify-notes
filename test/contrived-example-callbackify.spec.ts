import { getMsgAfterRandomSecsCallback } from '../lib/contrived-example-callbackify';

describe('callbackify getMsgAfterRandomSecsAsync', () => {
  const num1 = 1;
  const num2 = 4;
  const badInput = 'a';
  const patt = new RegExp(/waited for \d second\(s\)/);

  it('takes a callback that will log a message', (done) => {
    getMsgAfterRandomSecsCallback(num1, num2, (err: Error, msg: string) => {
      expect(msg).toMatch(patt);
      done();
    });

    expect.assertions(1);
  });

  it('errors if something is wrong', (done) => {
    jest.spyOn(global, 'setTimeout');
    // @ts-ignore - bad input on purpose
    getMsgAfterRandomSecsCallback(badInput, num2, (err: Error, msg: string) => {
      expect(err).toBeInstanceOf(TypeError);
      expect(err).toMatchSnapshot();
      expect(setTimeout).toHaveBeenCalledTimes(1);
      expect(setTimeout).toHaveBeenCalledWith(
        expect.any(Function),
        0,
      );
      done();
    });

    expect.assertions(4);
  });
});
