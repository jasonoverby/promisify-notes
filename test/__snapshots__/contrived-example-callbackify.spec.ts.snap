// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callbackify getMsgAfterRandomSecsAsync errors if something is wrong 1`] = `[TypeError: "One or both of these arguments is not a number: a, 4]`;
